segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java

```diff
@@ -172,6 +172,16 @@ public SegmentMetadata setOwnershipChanged(boolean value) {
         return setFlag(StatusFlags.OWNERSHIP_CHANGED, value);
     }
 
+    /**
+     * Sets whether writes should be atomic.
+     * @param value Value to set.
+     * @return This instance so that these calls can be chained.
+     */
+    public SegmentMetadata setAtomicWrites(boolean value) {
+        return setFlag(StatusFlags.ATOMIC_WRITES, value);
```

Comment:
Based on my conversation with @RaulGracia and this comment: https://github.com/pravega/pravega/pull/6827#discussion_r975394542
I am not sure this is something we ever want off. Because it would seem that while we gain a minor performance gain in the steady state, the other PR we would need to add would cost us quite a bit during failover.

Reply:
Atomicity of write operation in SLTs comes with the fine print and an exception.
As described [here](https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MetadataTransaction.java#L55-L57) Lazy commit functionality requires reconciliation. That is a clearly expected assumption and stated assumption for turning it on.

**How it works**
- With this optimization turned on, SLTS will not commit metadata changes if only change to metadata is an update to the length of segment + existing last chunk during append. In such cases updated metadata is kept in memory. 
- In all other cases metadata is always committed to table segment and up to date.
- In normal operation SLTS can guarantee atomic writes in all cases. Either whole data is appended and visible or none.
- In case of failure over, updated metadata in memory is lost and metadata in table segment is stale. So we need to reconcile and adjust length of last chunk, which happens in [claimOnwership](https://github.com/pravega/pravega/blob/01903b9ee88c8845a242b37bc565f2b1b7b32ca5/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java#L310) . We simply look at the current length of last chunk on LTS and adjust its metadata.  This reconciliation happens during [first openWrite after failure](https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java#L255-L258)
- [Similar logic](https://github.com/pravega/pravega/blob/01903b9ee88c8845a242b37bc565f2b1b7b32ca5/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java#L975) is implement while re-playing journal entries and recreating state of storage and container metadata segments.  
- A strong assumption in this recovery procedure is that upper layers are able to deals with partially written data and smartly handle it.
- This  works wonderfully for normal segments.

**What doesn't work**
In case of fail over SLTS breaks atomic update behavior ..in this one odd case you'll find partial data on LTS and that data will be reconciled.

**What is the problem?**

- For the reasons I am not totally convinced about the table segment BTree index logic makes an incorrect assumption that   SLTS writes are always atomic - so either whole page is written or none at all. Even in cases of failure it assumes that page was written entirely. It blindly assumes that last 12 bytes of segment always contain header.
- The big issue here is that btree index segment does not have any reconciliation logic whatsoever and it does not store any state in BK and can not recreate same page with same data again. 
- This is the real problem. Btree index logic needs to be fixed. Period.

**Where should we fix it?**
1. We can fix the atomicity in SLTS 
2. Or fix it in BTree by implementing reconciliation logic
My preference is that we fix the Btree logic and not mess with SLTS optimization

**Why new type of journal entry**
- With lazy commit turned off, the SLTS journal needs a record about each append operation  
- Originally we assumed that upper layers can handle partial writes , so never cared to make write operation 100% atomic even after failure. 
- Seems like if we don't want to fix Btree then we are left with no choice and are forced to implement atomicity in SLTS 100% of times.


I hope that answers your question.
My first preference is to fix Btree code instead of this  change.

Reply:
Note that this is an optimization that is relevant only in append mode. In no append mode it is always turned off. And because chunks are immutable and we add new chunk for each write in no-append mode, we don't have to deal with partial writes. 
In case of fail-over, [chunk creation] https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorage.java#L112-L123 will never succeed in the first place

Reply:
@tkaitchuck Another way to answer your question is to look at this flag as way to handle migration from data that was written by lazy-commit mode to code that has lazy commit mode turned off.

Note that currently lazy commit is on by default, So all data today is written assuming that after restart the new instance will check the length and 

When lazy commit is turned off we always commit metadata and there is no reason reconcile length of metadata based on length of LTS. (because in SLTS metadata was updated only after write was written successfully and fully)

So this flag is a marker flag in two sense
- when this flag is off then it indicates the data was written expecting length reconciliation. Any data written this change needs reconciliation (aka length adjustment) to work correctly. Therefore this change will reconcile length for any data written before this version
- when this flag is on then lazy commit can be never used with this segment again. And indicates that no length reconciliation is needed  any more.

This flag helps us migrate correctly from old version to new.

Reply:
Remove the option to configure lazy commit. From now on all segments are always written atomically.